src/main/java/it/smartcommunitylab/aac/saml/provider/SamlAuthenticationProvider.java

```diff
@@ -203,6 +236,32 @@ public Authentication doAuthenticate(Authentication authentication) throws Authe
         }
     }
 
+    private static List<Object> extractAcrAssertion(Response response) {
+        List<Object> acrValues = new ArrayList<>();
+        if ((response == null) || (response.getAssertions() == null) || (response.getAssertions().isEmpty())) {
+            return acrValues;
+        }
+        for (Assertion assertion : response.getAssertions()) {
+            for (Statement statement : assertion.getStatements()) {
+                if (!(statement instanceof AuthnStatementImpl)) {
+                    continue;
+                }
+                AuthnContext authnContext = ((AuthnStatementImpl) statement).getAuthnContext();
```

Comment:
refactor to stream and pick the first assertion with a non null authContext + acr